src/render/painter.js

```diff
@@ -249,9 +249,9 @@ class Painter {
         this.rasterBoundsSegments = SegmentVector.simpleSegment(0, 0, 4, 2);
 
         const viewportArray = new PosArray();
-        viewportArray.emplaceBack(0, 0);
-        viewportArray.emplaceBack(1, 0);
-        viewportArray.emplaceBack(0, 1);
+        viewportArray.emplaceBack(-1, -1);
+        viewportArray.emplaceBack(1, -1);
+        viewportArray.emplaceBack(-1, 1);
         viewportArray.emplaceBack(1, 1);
```

Comment:
`[[-4, -4], [4, -4], [0, 4]]` (a.k.a. [a-big-triangle](https://www.npmjs.com/package/a-big-triangle)) is also a nice way to cover the whole viewport. Not sure if it's still a caching win, but it's possibly-faster, equivalent, and less code, so I always advocate for it 😄

Reply:
You read my mind :) I actually also tried this yesterday and profiled it in this branch https://github.com/mapbox/mapbox-gl-js/tree/karim/heatmap-single-tri.

On my machine (Iris Plus Graphics 655) it stayed at 750us of GPU time with both approaches. So I then tried on a discrete GPU to be sure, and it was 100us with both approaches as well. My theory is that this shader isn't expensive enough to have any repercussion on its render time with a single or two triangles. This would work well when a lot of expensive fragment processing happens on the screen diagonal, where they might be evaluated twice (or even more with AA). But that's a good suggestion and a thing to keep in mind as we introduce more of these types of renders though.

Reply:
Yeah, that's why I usually lead with: "It's a performance win" and follow quickly with "well yeah, probably not an _observable_ win, but you can't argue that it's not still less code. 😛"